查询是否可以生成计算列

<!-- DON'T CHANGE THE TEMPLATE -->

## 查询字段或where条件，是否可以多个字段拼接组合

如以下sql
```sql
SELECT c.code as bcode,c.cdate,c.status,a.*,(a.sto_qty - a.dist_qty - a.allot_qty + a.retn_qty - a.allot_qty - a.loss_qty - a.borrow_qty + a.retn_qty2) as lft_qty,b.spec,b.type,b.unit,b.mate 
FROM s3_rec_item a
LEFT JOIN s3_cat_item b ON a.item_id = b.id
LEFT JOIN s3_rec_bill c ON a.rec_id = c.id
WHERE c.`status` = 3 
AND (a.sto_qty - a.dist_qty - a.allot_qty + a.retn_qty - a.allot_qty - a.loss_qty - a.borrow_qty + a.retn_qty2) > 0
```
问题1：字段
```sql
(a.sto_qty - a.dist_qty - a.allot_qty + a.retn_qty - a.allot_qty - a.loss_qty - a.borrow_qty + a.retn_qty2) as lft_qty
```
如何写

问题2：查询条件
```sql
(a.sto_qty - a.dist_qty - a.allot_qty + a.retn_qty - a.allot_qty - a.loss_qty - a.borrow_qty + a.retn_qty2) > 0
```
如何写

## Motivation

<!-- Describe the motivation behind it -->

## Related Issues

<!-- Link related issues here -->

编译没有问题，运行错误。
代码如下：
	calField := qryItem.StoQty.SubCol(qryItem.DistQty.SubCol(qryItem.AllotQty.AddCol(qryItem.RetnQty.SubCol(qryItem.LossQty.SubCol(qryItem.BorrowQty.AddCol(qryItem.RetnQty2))))))
	calFieldX, ok := calField.(field.Float64) //类型转换
	fmt.Println("类型转换是否成功：", ok)
	action := qryItemDo.Select(qryItem.ALL, qryBill.Code.As("bcode"), qryBill.Cdate, qryBill.Status,
		catItem.Mate, catItem.Spec, catItem.Type, catItem.Unit,
		calField.As("lft_qty")).
		LeftJoin(qryBill, qryBill.ID.EqCol(qryItem.RecID)).
		LeftJoin(catItem, catItem.ID.EqCol(qryItem.ItemID)).
		Where(calFieldX.Gt(0))
执行结果：
类型转换是否成功： false

2022/07/27 15:06:01 /Users/zhushuyan/go/pkg/mod/gorm.io/gen@v0.3.13/do.go:784 Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '` > ?' at line 1
[3.210ms] [rows:0] SELECT count(*) FROM `s3_rec_item` LEFT JOIN `s3_rec_bill` ON `s3_rec_bill`.`id` = `s3_rec_item`.`rec_id` LEFT JOIN `s3_cat_item` ON `s3_cat_item`.`id` = `s3_rec_item`.`item_id` WHERE ` > 0.000000
Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '` > ?' at line 1

类型转换不成功~~~~

类型强转：
`Where(field.Int64(calField).Gt(0))`
编译错误：
`
biz/s3_rec.go:97:21: cannot convert calField (variable of type field.Expr) to type field.Int64:
        need type assertion
`

代码：
```go
	qryBill, _ := query.S3RecBillDo()
	qryItem, qryItemDo := query.S3RecItemDo()
	catItem, catItemDo := query.S3CatItemDo()
	fmt.Println(catItemDo)
	calField := qryItem.StoQty.SubCol(qryItem.DistQty.SubCol(qryItem.AllotQty.AddCol(qryItem.RetnQty.SubCol(qryItem.LossQty.SubCol(qryItem.BorrowQty.AddCol(qryItem.RetnQty2)))))) //calField类型 filed.Expr
	action := qryItemDo.Select(qryItem.ALL, qryBill.Code.As("bcode"), qryBill.Cdate, qryBill.Status,
		catItem.Mate, catItem.Spec, catItem.Type, catItem.Unit,
		calField.As("lft_qty")). //设置别名
		LeftJoin(qryBill, qryBill.ID.EqCol(qryItem.RecID)).
		LeftJoin(catItem, catItem.ID.EqCol(qryItem.ItemID)).
		Where(field.Float64(calField).Gt(0)) //字段转换
```
编译：
```shell
biz/s3_rec.go:97:23: cannot convert calField (variable of type field.Expr) to type field.Float64:
        need type assertion
2022/07/27 15:38:51 ERROR    ▶ 0161 Failed to build the application: # www.hn24365.com/s3/biz
biz/s3_rec.go:97:23: cannot convert calField (variable of type field.Expr) to type field.Float64:
        need type assertion
```

--- a/field/export_test.go
+++ b/field/export_test.go
@@ -103,6 +103,14 @@ func TestExpr_Build(t *testing.T) {
 			Result: "`t1`.`id` + `t2`.`num`",
 		},
 		{
+			Expr:   field.NewInt("t1", "id").AddCol(field.NewInt("t1", "num")).SubCol(field.NewInt("t1", "age")),
+			Result: "`t1`.`id` + `t1`.`num` - `t1`.`age`",
+		},
+		{
+			Expr:   field.NewInt("t1", "id").AddCol(field.NewInt("t1", "num")).SubCol(field.NewInt("t1", "age")).MulCol(field.NewInt("t1", "age")).DivCol(field.NewInt("t1", "base")),
+			Result: "((`t1`.`id` + `t1`.`num` - `t1`.`age`) * (`t1`.`age`)) / (`t1`.`base`)",
+		},
+		{
 			Expr:         field.NewInt("t1", "id").AddCol(field.NewInt("t2", "num").Add(1)),
 			Result:       "`t1`.`id` + `t2`.`num`+?",
 			ExpectedVars: []interface{}{int(1)},

--- a/field/expr.go
+++ b/field/expr.go
@@ -25,6 +25,12 @@ type Expr interface {
 	BuildWithArgs(*gorm.Statement) (query sql, args []interface{})
 	RawExpr() expression
 
+	// col operate expression
+	AddCol(col Expr) Expr
+	SubCol(col Expr) Expr
+	MulCol(col Expr) Expr
+	DivCol(col Expr) Expr
+
 	// implement Condition
 	BeCond() interface{}
 	CondError() error
@@ -221,6 +227,14 @@ func (e expr) AddCol(col Expr) Expr {
 
 func (e expr) SubCol(col Expr) Expr {
 	return e.setE(clause.Expr{SQL: "? - ?", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})
+}
+
+func (e expr) MulCol(col Expr) Expr {
+	return e.setE(clause.Expr{SQL: "(?) * (?)", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})
+}
+
+func (e expr) DivCol(col Expr) Expr {
+	return e.setE(clause.Expr{SQL: "(?) / (?)", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})
 }
 
 // ======================== keyword ========================

--- a/field/expr.go
+++ b/field/expr.go
@@ -222,19 +222,19 @@ func (e expr) SetCol(col Expr) AssignExpr {
 
 // ======================== operate columns ========================
 func (e expr) AddCol(col Expr) Expr {
-	return e.setE(clause.Expr{SQL: "? + ?", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})
+	return Field{e.setE(clause.Expr{SQL: "? + ?", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})}
 }
 
 func (e expr) SubCol(col Expr) Expr {
-	return e.setE(clause.Expr{SQL: "? - ?", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})
+	return Field{e.setE(clause.Expr{SQL: "? - ?", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})}
 }
 
 func (e expr) MulCol(col Expr) Expr {
-	return e.setE(clause.Expr{SQL: "(?) * (?)", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})
+	return Field{e.setE(clause.Expr{SQL: "(?) * (?)", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})}
 }
 
 func (e expr) DivCol(col Expr) Expr {
-	return e.setE(clause.Expr{SQL: "(?) / (?)", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})
+	return Field{e.setE(clause.Expr{SQL: "(?) / (?)", Vars: []interface{}{e.RawExpr(), col.RawExpr()}})}
 }
 
 // ======================== keyword ========================

